### PR TITLE
Add special ops for BatchNorm symbolic differentiation

### PR DESCRIPTION
**Pull request by apaszke**
The main problem there is with differentiating batch norm statically
is that we make a lot of complex run-time decisions about the backend
we choose. Then, the autograd derivatives are implemented for every
backend separately, which makes sense, because they might be saving
buffers containing different values. To resolve the issue, the forward
op returns an index of the chosen backend, and the backward function
takes it as an argument, such that it knows how to interpret the buffers.

### CODE REVIEW

**Review by ailzhang — APPROVED** [review]
LGTM! I only checked the AD part tho. 
Test failures seem unrelated...

**Comment by zou3519** [comment]
github ui says this needs a rebase

**Comment by apaszke** [comment]
Two failures are build timeouts, one is distributed test flakiness.

File: torch/csrc/jit/symbolic_script.cpp [hunk 1]
```diff
@@ -20,6 +20,31 @@ const std::vector<std::string> functions = {
...
+                       running_var : Optional[Tensor],
+                       training : bool,
+                       momentum : float,
+                       eps : float,
+                       cudnn_enabled : bool):
+
+            output, save1, save2, impl_idx = torch._batch_norm_impl_index(
+                input, weight, bias, running_mean, running_var, training,
+                momentum, eps, cudnn_enabled)
+            has_weight = weight._defined()
```
Comment (zdevito) [inline torch/csrc/jit/symbolic_script.cpp]: These should be `x is not None`. If that doesn't work then we need to file an issue.

Comment (apaszke) [reply torch/csrc/jit/symbolic_script.cpp]: Oh I had no idea I could do that. Will update.

**Review by zdevito — APPROVED** [review]
